tests/components/vesync/test_init.py

```diff
@@ -67,37 +71,115 @@ async def test_async_setup_entry__no_devices(
 
     assert manager.login.call_count == 1
     assert hass.data[DOMAIN][VS_MANAGER] == manager
-    assert not hass.data[DOMAIN][VS_SWITCHES]
     assert not hass.data[DOMAIN][VS_FANS]
     assert not hass.data[DOMAIN][VS_LIGHTS]
     assert not hass.data[DOMAIN][VS_SENSORS]
+    assert not hass.data[DOMAIN][VS_SWITCHES]
 
 
-async def test_async_setup_entry__loads_fans(
-    hass: HomeAssistant, config_entry: ConfigEntry, manager: VeSync, fan
+async def test_async_setup_entry__with_devices(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    manager_devices: VeSync,
 ) -> None:
     """Test setup connects to vesync and loads fan platform."""
-    fans = [fan]
-    manager.fans = fans
-    manager._dev_list = {
-        "fans": fans,
-    }
-
     with patch.object(
+        hass,
+        "async_add_executor_job",
+        new=AsyncMock(),
+    ) as mock_add_executor_job, patch.object(
         hass.config_entries, "async_forward_entry_setups"
     ) as setups_mock, patch.object(
-        hass.config_entries, "async_forward_entry_setup"
-    ) as setup_mock:
+        hass.services, "async_register"
+    ) as register_mock:
         assert await async_setup_entry(hass, config_entry)
         # Assert platforms loaded
         await hass.async_block_till_done()
+
+        assert mock_add_executor_job.call_count == 2
+        assert mock_add_executor_job.call_args_list == [
+            call(manager_devices.login),
+            call(manager_devices.update),
+        ]
         assert setups_mock.call_count == 1
         assert setups_mock.call_args.args[0] == config_entry
-        assert setups_mock.call_args.args[1] == [Platform.FAN, Platform.SENSOR]
-        assert setup_mock.call_count == 0
-    assert manager.login.call_count == 1
-    assert hass.data[DOMAIN][VS_MANAGER] == manager
-    assert not hass.data[DOMAIN][VS_SWITCHES]
-    assert hass.data[DOMAIN][VS_FANS] == [fan]
-    assert not hass.data[DOMAIN][VS_LIGHTS]
-    assert hass.data[DOMAIN][VS_SENSORS] == [fan]
+        assert list(setups_mock.call_args.args[1]) == [
+            Platform.SWITCH,
+            Platform.FAN,
+            Platform.LIGHT,
+            Platform.SENSOR,
+        ]
+        assert register_mock.call_count == 1
+        assert register_mock.call_args.args[0] == DOMAIN
+        assert register_mock.call_args.args[1] == SERVICE_UPDATE_DEVS
+        assert callable(register_mock.call_args.args[2])
+
+    assert hass.data[DOMAIN][VS_MANAGER] == manager_devices
+    assert len(hass.data[DOMAIN][VS_FANS]) == 1
+    assert len(hass.data[DOMAIN][VS_LIGHTS]) == 2
+    assert len(hass.data[DOMAIN][VS_SENSORS]) == 2
+    assert len(hass.data[DOMAIN][VS_SWITCHES]) == 2
+
+
+async def test_asynch_setup_entry__loaded_state(
+    hass: HomeAssistant,
+    setup_platform,
+    snapshot: SnapshotAssertion,
+) -> None:
+    """Test the resulting setup state is as expected."""
```

Comment:
This would be better handled inside platform tests.
You can patch `PLATFORMS` dict, and you can combine that with `parametrize` and `snapshot` to show expected snapshots for each individual device.

See an example with two device types on sfr_box:
 https://github.com/home-assistant/core/blob/6e23e00b5a6152bec7ddd6d1fbd851ee552c84aa/tests/components/sfr_box/test_binary_sensor.py#L19-L23

You can add a test for a single platform to ensure that we are on the same wave-length, then add the tests for the other platforms as a follow-up PR.